### PR TITLE
Fixed bug in the post processing of detect_c2c3

### LINKED ISSUE #2191: set_label_vertebrae fails in version 4.0.0-beta.1
The command sct_label_vertabrae of SCT version 4.0.0-beta.1 failed to find C2-C3 in a T1w image. In contrast, for the same subject, the SCT version 3.2.7 identified automatically C2-C3.

In attachment, follows the description of the command used and messages obtained. In addition, follows a test imaging. 

SCT version: 3.2.7

<details>

~~~
evans:~/Documents/testEnigma_2 $ sct_label_vertebrae -i cont_03.nii.gz -s cont_03_seg.nii.gz -c t1

--
Spinal Cord Toolbox (version: 3.2.7)
Running /Users/thiago_admin/sct_3.2.7/scripts/sct_label_vertebrae.py -i cont_03.nii.gz -s cont_03_seg.nii.gz -c t1

Create temporary folder (/var/folders/dv/9t9_nj0d6ln20k7bw4wlqtvr0000gv/T/sct-20190313104129.360303-label_vertebrae-lphp2t)...

Copying input data to tmp folder...

Create label to identify disc...

Straighten spinal cord...
sct_straighten_spinalcord -i data.nii -s segmentation.nii.gz -r 1 # in /private/var/folders/dv/9t9_nj0d6ln20k7bw4wlqtvr0000gv/T/sct-20190313104129.360303-label_vertebrae-lphp2t

Resample to 0.5mm isotropic...
sct_resample -i data_straight.nii -mm 0.5x0.5x0.5 -x linear -o data_straightr.nii # in /private/var/folders/dv/9t9_nj0d6ln20k7bw4wlqtvr0000gv/T/sct-20190313104129.360303-label_vertebrae-lphp2t

Apply straightening to segmentation...
sct_apply_transfo -i segmentation.nii.gz -d data_straightr.nii -w warp_curve2straight.nii.gz -o segmentation_straight.nii.gz -x linear # in /private/var/folders/dv/9t9_nj0d6ln20k7bw4wlqtvr0000gv/T/sct-20190313104129.360303-label_vertebrae-lphp2t
sct_maths -i segmentation_straight.nii.gz -thr 0.5 -o segmentation_straight.nii.gz # in /private/var/folders/dv/9t9_nj0d6ln20k7bw4wlqtvr0000gv/T/sct-20190313104129.360303-label_vertebrae-lphp2t

Look for template...
Path template: /Users/thiago_admin/sct_3.2.7/data/PAM50

Open template and vertebral levels...

Disc values from template: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
Z-values for each disc: [962, 939, 908, 871, 834, 801, 770, 736, 693, 647, 601, 552, 501, 450, 397, 343, 290, 232, 169, 106, 56]
Distances between discs (in voxel): [23, 31, 37, 37, 33, 31, 34, 43, 46, 46, 49, 51, 51, 53, 54, 53, 58, 63, 63, 50]

Detect C2/C3 disk...
.. Peak found: z=93 (correlation = 0.374294261593)

Detect intervertebral discs...
Current disc: 2 (z=78). Direction: superior
.. Peak found: z=0 (correlation = 0.52335737801)
Current disc: 1 (z=109). Direction: superior
.. correcting factor: 1.0
.. Switching to inferior direction.
Current disc: 3 (z=41). Direction: inferior
.. Peak found: z=9 (correlation = 0.519620547701)
.. correcting factor: 0.878378378378
Current disc: 4 (z=18). Direction: inferior
.. Peak found: z=3 (correlation = 0.598587032476)
.. correcting factor: 0.846846846847
Adding top disc based on adjusted template distance: #0
.. approximate distance: 19

Un-straighten labeling...
sct_apply_transfo -i segmentation_straight_labeled.nii.gz -d segmentation.nii.gz -w warp_straight2curve.nii.gz -o segmentation_labeled.nii.gz -x nn # in /private/var/folders/dv/9t9_nj0d6ln20k7bw4wlqtvr0000gv/T/sct-20190313104129.360303-label_vertebrae-lphp2t

Clean labeled segmentation (correct interpolation errors)...
sct_maths -i segmentation_labeled.nii.gz -mul segmentation.nii.gz -o segmentation_labeled_mul.nii.gz # in /private/var/folders/dv/9t9_nj0d6ln20k7bw4wlqtvr0000gv/T/sct-20190313104129.360303-label_vertebrae-lphp2t
sct_maths -i segmentation_labeled.nii.gz -dilate 2 -o segmentation_labeled_dilate.nii.gz # in /private/var/folders/dv/9t9_nj0d6ln20k7bw4wlqtvr0000gv/T/sct-20190313104129.360303-label_vertebrae-lphp2t
sct_maths -i segmentation_labeled_mul.nii.gz -bin 0 -o segmentation_labeled_mul_bin.nii.gz # in /private/var/folders/dv/9t9_nj0d6ln20k7bw4wlqtvr0000gv/T/sct-20190313104129.360303-label_vertebrae-lphp2t
WARNING: File /private/var/folders/dv/9t9_nj0d6ln20k7bw4wlqtvr0000gv/T/sct-20190313104129.360303-label_vertebrae-lphp2t/segmentation_labeled.nii.gz already exists. Will overwrite it.

Label discs...

Generate output files...
  File created: ./cont_03_seg_labeled.nii.gz
  File created: ./cont_03_seg_labeled_discs.nii.gz
  File created: ./warp_curve2straight.nii.gz
  File created: ./warp_straight2curve.nii.gz
  File created: ./straight_ref.nii.gz

Remove temporary files...
rm -rf /var/folders/dv/9t9_nj0d6ln20k7bw4wlqtvr0000gv/T/sct-20190313104129.360303-label_vertebrae-lphp2t

Done! To view results, type:
fslview /Users/thiago_admin/Documents/testEnigma_2/cont_03.nii.gz -t 1 ./cont_03_seg_labeled.nii.gz -l MGH-Subcortical -t 0.5 &

Total processing time: 1 min 31 s
~~~

</details>

VetebraeLabel_v327

SCT version: 4.b.1

<details>

~~~
evans:~/Documents/testEnigma_3 $ sct_label_vertebrae -i cont_03.nii.gz -s cont_03_seg.nii.gz -c t1

--
Spinal Cord Toolbox (version: 4.0.0-beta.1)
Running /Users/thiago_admin/sct_4.0.0-beta.1/scripts/sct_label_vertebrae.py -i cont_03.nii.gz -s cont_03_seg.nii.gz -c t1

Create temporary folder (/var/folders/dv/9t9_nj0d6ln20k7bw4wlqtvr0000gv/T/sct-20190313115830.590073-label_vertebrae-SoameE)...

Copying input data to tmp folder...

Create label to identify disc...
Creating temporary folder...
Run C2-C3 detector...
C2-C3 not detected...
Remove temporary files...
rm -rf /var/folders/dv/9t9_nj0d6ln20k7bw4wlqtvr0000gv/T/sct-20190313115833.209371-oAvLK1
Automatic C2-C3 detection failed. Please provide manual label with sct_label_utils
Traceback (most recent call last):
  File "/Users/thiago_admin/sct_4.0.0-beta.1/scripts/sct_label_vertebrae.py", line 842, in <module>
    main()
  File "/Users/thiago_admin/sct_4.0.0-beta.1/scripts/sct_label_vertebrae.py", line 256, in main
    sct.printv('Automatic C2-C3 detection failed. Please provide manual label with sct_label_utils', 1, 'error')
  File "/Users/thiago_admin/sct_4.0.0-beta.1/scripts/sct_utils.py", line 1016, in printv
    raise RuntimeError("printv(..., type=\"error\")")
RuntimeError: printv(..., type="error")
Sentry is attempting to send 1 pending error messages
Waiting up to 10 seconds
Press Ctrl-C to quit
Total processing time: 0 min 6 s
~~~

</details>

[SCT_Label_Vertebrae.docx]

(https://github.com/neuropoly/spinalcordtoolbox/files/2966797/SCT_Label_Vertebrae.docx)
[cont_03.nii.gz](https://github.com/neuropoly/spinalcordtoolbox/files/2966798/cont_03.nii.gz)

### PR DESCRIPTION
**Pull request by charleygros**
This PR aims at:
- Fixing a bug in the post processing of `detect_c2c3` module. Fixes #2191 
- Allowing a way not to remove temporary files, in order to facilitate the debugging.

### Post processing bug:
We create a mask around the centerline, which is used in the post-processing of `detect_c2c3` module. In the new implementation, we only consider z-slices where the spinal cord is present (ie not in the brain area to speed up the process) and do not forget to iterate along the z-axis (which was not the case in the former implementation.. shame.).

### Temporary files:
Keep temporary files of `detect_c2c3` module is now possible.
To test it:
```
cd spinalcordtoolbox/spinalcordtoolbox/vertebrae/
python
```
Then,
```
>>> from detect_c2c3 import detect_c2c3_from_file
>>> detect_c2c3_from_file('/home/charley/data/thiago/cont_03.nii.gz', '/home/charley/data/thiago/cont_03_seg.nii.gz', 't1', fname_c2c3='/home/charley/data/thiago/cont_03_c2c3.nii.gz', verbose=2)
```

And:
```
cd /home/charley/data/thiago/
fsleyes tmp_midSlice.nii.gz tmp_midSlice_mask.nii.gz -cm red tmp_midSlice_before_postPro.nii.gz -cm blue tmp_midSlice_after_postPro.nii.gz -cm green
```

### CODE REVIEW

File: spinalcordtoolbox/vertebrae/detect_c2c3.py [hunk 1]
```diff
@@ -30,12 +30,13 @@
...
 
 
-def detect_c2c3(nii_im, nii_seg, contrast, nb_sag_avg=7.0, verbose=1):
+def detect_c2c3(nii_im, nii_seg, contrast, nb_sag_avg=7.0, remove_temp_files=1, verbose=1):
     """
     Detect the posterior edge of C2-C3 disc.
     :param nii_im:
     :param nii_seg:
     :param contrast:
+    :parem remove_temp_file:
```
Comment (jcohenadad) [inline spinalcordtoolbox/vertebrae/detect_c2c3.py]: small typo

File: spinalcordtoolbox/vertebrae/detect_c2c3.py [hunk 2]
```diff
@@ -74,39 +74,35 @@ def detect_c2c3(nii_im, nii_seg, contrast, nb_sag_avg=7.0, verbose=1):
     sct.run(cmd_detection, verbose=0, raise_exception=False)
 
     pred = nib.load('data_midSlice_pred_svm.hdr').get_data()
+    # save the prediction data before post-processing
```
Comment (jcohenadad) [inline spinalcordtoolbox/vertebrae/detect_c2c3.py]: the code below does not save anything --> maybe rephrase the comment?

File: spinalcordtoolbox/vertebrae/detect_c2c3.py [hunk 3]
```diff
@@ -74,39 +74,35 @@ def detect_c2c3(nii_im, nii_seg, contrast, nb_sag_avg=7.0, verbose=1):
...
+    nii_pred_before_postPro = nii_midSlice.copy()
+    nii_pred_before_postPro.data = pred
 
     # Create mask along centerline
     midSlice_mask = np.zeros(midSlice_seg.shape)
     mask_halfSize = int(np.rint(25.0 / nii_midSlice.dim[4]))
     for z in range(midSlice_mask.shape[1]):
```
Comment (jcohenadad) [inline spinalcordtoolbox/vertebrae/detect_c2c3.py]: usually we operate in RPI orientation, meaning that S-I axis would correspond to .shape[2] (not .shape[1]). Could you add comments that would help understanding the code (e.g. what orientation is assumed here)

File: spinalcordtoolbox/vertebrae/detect_c2c3.py [hunk 4]
```diff
@@ -74,39 +74,35 @@ def detect_c2c3(nii_im, nii_seg, contrast, nb_sag_avg=7.0, verbose=1):
...
         row = midSlice_seg[:, z]
-        if np.any(row):
-            med_y = int(np.rint(np.median(np.where(row))))
-            midSlice_mask[med_y-mask_halfSize:med_y+mask_halfSize] = 1
+        if np.any(row > 0):
+            med_y = int(np.rint(np.median(np.where(row > 0))))
+            midSlice_mask[med_y-mask_halfSize:med_y+mask_halfSize, z] = 1
+    # save the created mask
```
Comment (jcohenadad) [inline spinalcordtoolbox/vertebrae/detect_c2c3.py]: same as previous comment-- nothing seems to be saved here

File: spinalcordtoolbox/vertebrae/detect_c2c3.py [hunk 5]
```diff
@@ -74,39 +74,35 @@ def detect_c2c3(nii_im, nii_seg, contrast, nb_sag_avg=7.0, verbose=1):
...
+    nii_postPro_mask = nii_midSlice.copy()
+    nii_postPro_mask.data = midSlice_mask
 
     # mask prediction
     pred[midSlice_mask == 0] = 0
-    # dist_medulla = 30.0 if contrast == 't1' else 40.0  # Observation: segmentation ends higher on t2 images than t1
-    # z_seg_max -= int(np.rint(dist_medulla / nii_midSlice.dim[5])) # TODO: take into account the curvature
     pred[:, z_seg_max:] = 0  # Mask above SC segmentation
+    # save the prediction data after post-processing
```
Comment (jcohenadad) [inline spinalcordtoolbox/vertebrae/detect_c2c3.py]: same as previous comment-- nothing seems to be saved here

File: spinalcordtoolbox/vertebrae/detect_c2c3.py [hunk 6]
```diff
@@ -139,8 +138,21 @@ def detect_c2c3_from_file(fname_im, fname_seg, contrast, fname_c2c3=None, verbos
...
-    nii_c2c3 = detect_c2c3(nii_im, nii_seg, contrast, verbose=verbose)
+    if remove_temp_files:
+        nii_c2c3 = detect_c2c3(nii_im.copy(), nii_seg, contrast, verbose=verbose)
+    else:
+        nii_c2c3, nii_midSlice, nii_mask, nii_after_postPro, nii_before_postPro = detect_c2c3(nii_im.copy(),
+                                                                                                nii_seg,
+                                                                                                contrast,
+                                                                                                remove_temp_files=remove_temp_files,
+                                                                                                verbose=verbose)
+        nii_midSlice.save(os.path.join(os.path.dirname(nii_im.absolutepath), "tmp_midSlice.nii.gz"))
```
Comment (jcohenadad) [inline spinalcordtoolbox/vertebrae/detect_c2c3.py]: now that's a bit confusing. Here we explicitly save debugging file by setting "remove_temp_files" to 0, which is counter-intuitive. 

the variable `remove_temp_files` should only be used for the action "remove temp file".

if we want to save debugging files, we should use another variable (e.g. debugging), or, as mostly done in SCT, use verbose=2 mode.

**Comment by charleygros** [comment]
@jcohenadad : Thank you for your feedback.
Ready to re-review!

File: spinalcordtoolbox/vertebrae/detect_c2c3.py [hunk 7]
```diff
@@ -121,7 +116,10 @@ def detect_c2c3(nii_im, nii_seg, contrast, nb_sag_avg=7.0, verbose=1):
     tmp_folder.cleanup()
 
     nii_c2c3.change_orientation(orientation_init)
-    return nii_c2c3
+    if verbose < 2:
```
Comment (jcohenadad) [inline spinalcordtoolbox/vertebrae/detect_c2c3.py]: why not simply returning everything, then in `detect_c2c3_from_file` you do:
~~~
nii_c2c3, nii_midSlice, nii_mask, nii_after_postPro, nii_before_postPro = detect_c2c3(nii_im.copy(),
                                                                                                nii_seg,
                                                                                                contrast,
                                                                                                verbose=verbose)
if verbose > 1:
    # save debugging files
    path_temp = sct.tmp_create
    nii_midSlice.save(os.path.join(os.path.dirname(nii_im.absolutepath), os.path.join(path_temp, "tmp_midSlice.nii.gz")))
   # etc.
~~~

also, good habit to use `sct.temp_dir`, to avoid overwriting, in case use likes to call his files "tmp_midSlice.nii.gz"

Comment (jcohenadad) [reply spinalcordtoolbox/vertebrae/detect_c2c3.py]: update: actually, it would be less confusing to save the debugging files inside the calling function (not in  detect_c2c3_from_file)

Comment (charleygros) [reply spinalcordtoolbox/vertebrae/detect_c2c3.py]: Good point!
Done
